File: swift/Sources/MossFormer2SS/CLayerNorm.swift
import Foundation
import MLX
import MLXNN

/// MLX implementation of CLayerNorm (Channel-wise Layer Normalization).
///
/// This class applies layer normalization along the channel dimension.
/// Unlike the PyTorch version which expects [N, C, T], this MLX version
/// works directly with [N, T, C] format to avoid unnecessary transpositions.
public class CLayerNorm: Module, UnaryLayer {
    public let normalizedShape: Int
    public let eps: Float
    public let elementwiseAffine: Bool
    
    public var weight: MLXArray?
    public var bias: MLXArray?
    
    /// Initialize CLayerNorm
    /// - Parameters:
    ///   - normalizedShape: Input shape from last dimension
    ///   - eps: Small value for numerical stability (default: 1e-8)
    ///   - elementwiseAffine: Whether to use learnable affine parameters (default: true)
    public init(
        normalizedShape: Int,
        eps: Float = 1e-8,
        elementwiseAffine: Bool = true
    ) {
        self.normalizedShape = normalizedShape
        self.eps = eps
        self.elementwiseAffine = elementwiseAffine
        
        if elementwiseAffine {
            self.weight = MLXArray.ones([normalizedShape])
            self.bias = MLXArray.zeros([normalizedShape])
        }
        
        super.init()
    }
    
    /// Forward pass applying channel-wise layer normalization
    /// - Parameter x: Input tensor of shape [batch_size, sequence_length, channels]
    /// - Returns: Normalized tensor of shape [batch_size, sequence_length, channels]
    public func callAsFunction(_ x: MLXArray) -> MLXArray {
        guard x.ndim == 3 else {
            fatalError("CLayerNorm only accepts 3-D tensor as input, got \(x.ndim)D")
        }
        
        // x is already in [N, T, C] format
        // Apply LayerNorm along the channel dimension (last dimension)
        let mean = x.mean(axis: -1, keepDims: true)
        let variance = x.variance(axis: -1, keepDims: true)
        var output = (x - mean) / MLX.sqrt(variance + eps)
        
        if elementwiseAffine, let weight = weight, let bias = bias {
            output = output * weight + bias
        }
        
        return output
    }
}

File: python/clayernorm.py
import mlx.core as mx
import mlx.nn as nn

class CLayerNorm_MLX(nn.Module):
    """
    MLX implementation of CLayerNorm (Channel-wise Layer Normalization).
    
    This class applies layer normalization along the channel dimension.
    Unlike the PyTorch version which expects [N, C, T], this MLX version
    works directly with [N, T, C] format to avoid unnecessary transpositions.
    
    Args:
        normalized_shape: Input shape from last dimension
        eps: Small value for numerical stability (default: 1e-8)
        elementwise_affine: Whether to use learnable affine parameters (default: True)
    
    Shape:
        - Input: [batch_size, sequence_length, channels]
        - Output: [batch_size, sequence_length, channels]
    """
    
    def __init__(self, normalized_shape, eps=1e-8, elementwise_affine=True):
        super().__init__()
        self.normalized_shape = normalized_shape
        self.eps = eps
        self.elementwise_affine = elementwise_affine
        
        if self.elementwise_affine:
            self.weight = mx.ones((normalized_shape,))
            self.bias = mx.zeros((normalized_shape,))
        else:
            self.weight = None
            self.bias = None
    
    def __call__(self, x):
        """Forward pass applying channel-wise layer normalization."""
        if x.ndim != 3:
            raise RuntimeError(f'CLayerNorm_MLX only accepts 3-D tensor as input, got {x.ndim}D')
        
        # x is already in [N, T, C] format
        # Apply LayerNorm along the channel dimension (last dimension)
        mean = mx.mean(x, axis=-1, keepdims=True)
        var = mx.var(x, axis=-1, keepdims=True)
        x = (x - mean) / mx.sqrt(var + self.eps)
        
        if self.elementwise_affine:
            x = x * self.weight + self.bias
        
        return x